### ISSUE TITLE
After changing file with notepad, caught in "the file changed" loop

### ISSUE THREAD

**Issue by pkellner** [RickStrahl/MarkdownMonster#45]
file changing in background is causing "file change detected" dialog to continually come up and not be able to get rid of.

**Comment by RickStrahl** [comment]
Hi Peter,

I can't repro this.

* Open document in MM
* Use Notepad to make a change
* Save document in Notepad
* Switch back to MM
* Get dialog to say file has changed
* Click Yes
* File changes
* No more dialog until I make another change

Can you check and see if the file date and size of the file on disk keeps changing on you?

The logic in code basically saves a CRC with the open document and after you say No or Yes on that dialog the CRC is updated to the image that lives on disk. The CRC shouldn't change unless the file's content is changing.

**Comment by ferventcoder** [comment]
I was just able to repro this. Not sure if it has to do with changing the encoding or not, but for me it was a UTF8 changed to a UTF 8 w/BOM, that I then changed back without closing MM. Granted I also had a document open for a very long time before I did this so there are some variables into what may have caused this.

**Comment by ferventcoder** [comment]
Enough clicks to yes, no and the "x" and it eventually closed off.

**Comment by ferventcoder** [comment]
I tried to capture a quick video of it to share but it didn't quite capture correctly.

**Comment by RickStrahl** [comment]
As I've been working on the Encoding stuff I've played around with this piece of it quite a bit today myself but I can't make it fail, even when changing the underlying encoding in Notepad (or sublime etc.). MM should now properly show the changed encoding.

Give another try in 1.0.10 and see if this works better - if encoding was an issue in this that should now work properly.

**Comment by pkellner** [comment]
Just updated to .11 and get same error on fresh win10 reboot. video showing here.  https://youtu.be/kJgG37Rn7So

**Comment by ferventcoder** [comment]
@pkellner 👍  that's what I've seen as well. Eventually it quits for me, but after a lot of clicks. Even clicking no doesn't remove it.

**Comment by RickStrahl** [comment]
Thanks for the Video Peter. 

Installing Notepad++ to see if there's something that it's doing that's causing the file change triggering to happen, but I can't imagine what it could be. Will take a look.

**Comment by RickStrahl** [comment]
Still trying to repo this.  Installed NP++ and followed the same steps you're doing in the video I still don't see this.

Are you guys running under Parallels or some other virtualization?  

Could it be a permissions thing? Trying to read/write a file out of a folder that doesn't have write access perhaps (unlikely but should ask)?

Just to clarify the way that this works in MM: 

Everytime a file is loaded or saved on disk MM captures a CRC of the file on disk. Then when you Activate the tab in the editor, MM checks the file's actual CRC against this saved CRC. If the CRCs don't match, then the dialog displays. In theory, this should work and from what I can see here on my setup it looks like it does.

So something in the environment appears to be screwing with this - and I'm trying to figure out what it could be.

I'm going to try running under Parallels and see what happens.

**Comment by ferventcoder** [comment]
Running under VMWare Fusion here.

**Comment by pkellner** [comment]
@RickStrahl running out the door, but just some quick answers:

Native Windows Machine (desktop)

everything from same folder

Dump a bunch of logging in?

Gone till Sunday.

**Comment by RickStrahl** [comment]
I've made a few small tweaks that check for load document errors. Maybe you can try them out with this interim build (1.0.12) from here:

https://github.com/RickStrahl/MarkdownMonster/tree/master/Install/Builds/CurrentRelease

Peter i noticed in your video that after to you press Yes the editor is not actually updating its content, which seems to suggest the file load might be failing. This code checks for that and pops up a separate dialog and updates the CRC explicitly and so should stop further issues. If you get that second dialog (load failed) though, that means something else is going on with the file (like the file is locked - perhaps by AV software? which could account for the failure).

**Comment by RickStrahl** [comment]
Any updates on this guys? 

Current version is 1.0.18 and there are  a few tweaks that might have helped with this...

**Comment by pkellner** [comment]
Have not been home much.  Will test tonight.  BTW, it was git updates in addition to notepad2 caused issues.

**Comment by ferventcoder** [comment]
Turkey weekend for me so I didn't get to dive back in.

**Comment by RickStrahl** [comment]
Sorry - didn't mean to be pushy :-) Just checking back. Let me know what you find once you update to a later version. 

Certainly want to get to the bottom of this.

**Comment by pkellner** [comment]
push away!  Just tested latest version with same result. I also installed it on my macbook parallels and background update of the doc loaded as expected (no repeating error).

**Comment by ferventcoder** [comment]
No worries, did not think you were being pushy. :)

**Comment by pkellner** [comment]
Hi Rick,
I did some debugging and I'm 90% sure I know what the problem is and wish I could tell you the solution (sadly, have to get back to other work at the moment).

In MainWindow.xaml.cs, what is happening is that you are calling MessageBox.Show after you detect a CRC changed.  After the user closes MessageBox by answering, MessageBox fires another windows activation event from it's calling window and triggers this method again (before you have a chance to call doc.UpdateCrc.  

Here is docs on that:  https://msdn.microsoft.com/en-us/library/system.windows.window.showdialog(v=vs.110).aspx

I may be wrong, but I think another strategy might be to register a file watcher and then, when the change event gets fired on that file watcher by the OS, prompt and update the dialog. I've not used that class in many years so it may not even work like that anymore.

Sorry I couldn't fix it :(

**Comment by pkellner** [comment]
BTW, I think the reason is it fails on mine is I have a screamin fast computer and graphics card and the windows close event is finishing faster then on slower computers (who knows though)

**Comment by RickStrahl** [comment]
Thanks Peter - you might be on to something here. Not sure I buy the fast machine thing - I'm also running on a fast box, but looking at the code there is a path through there that doesn't update the CRC explicitly and activation could play a role.

I made a change to always update the CRC immediately, which is the behavior we want anyway. The CRC check is to determine what's on disk and once notified of the change we don't want to be notified again for the same file. So moving the CRC update before the MBs should avoid future dialogs.

f5719dbad56fb6308bac51bc1a58047b9fe7cec1

I've updated the installer with an intermediate build which you can grab from the Git Repo:
https://github.com/RickStrahl/MarkdownMonster/blob/master/Install/Builds/CurrentRelease/MarkdownMonsterSetup.exe

**Comment by pkellner** [comment]
That worked.  I tested answering question yes and no and got expected results.

On my  machine, it was the update above the messagebox that fixed it. For some reason I thought that would not help but did not try.  When I debugged, as soon as messagebox returned answer, the next line of execution was the line a few lines above checking crc.  the why mine fails and other machines work is still a mystery to me. I would think they would all fail.

Thanks for fixing and MM.
